File: analysis/src/main/java/com/greencross/lims/entity/readonly/Work.java
package com.greencross.lims.entity.readonly;

import lombok.Builder;
import lombok.Data;
import lombok.Getter;
import lombok.experimental.Accessors;

import javax.persistence.*;
import java.io.Serializable;
import java.util.UUID;

@Entity
@Table(schema="public", name = "work")
@Getter
@Accessors(fluent = true)
public class Work {
	@EmbeddedId
	private WorkPK pk;
	@Column(name="serial", length=13)
	private String serial;
	@ManyToOne(fetch= FetchType.LAZY)
	@JoinColumns({
			@JoinColumn(name="sample", referencedColumnName="sample", insertable=false, updatable=false)
			, @JoinColumn(name="service", referencedColumnName="service", insertable=false, updatable=false)})
	private Request request;

	@Embeddable
	@Data
	@Accessors(fluent = true)
	public static class WorkPK implements Serializable {
		@Column(name="sheet", nullable=false, updatable=false)
		private UUID sheet;
		@Column(name="worklist", columnDefinition="integer", nullable=false, updatable=false)
		private Integer work;
		@Column(name="sample", nullable=false, updatable=false)
		private Long sample;
		@Column(name="service", length=8, nullable=false, updatable=false)
		private String service;

		public WorkPK() {}

		@Builder
		public WorkPK(UUID sheet, Integer work, Long sample, String service) {
			this.sheet = sheet;
			this.work = work;
			this.sample = sample;
			this.service = service;
		}
	}
}

File: analysis/src/main/kotlin/com/gcgenome/lims/entity/Work.kt
package com.gcgenome.lims.entity

import com.infobip.spring.data.jdbc.annotation.processor.Schema
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Transient
import org.springframework.data.domain.Persistable
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import java.util.*

@Schema("public")
@Table("work")
data class Work(
    @Column("sheet")        val sheet:      UUID,
    @Column("worklist")     val worklist:   Int,
    @Column("sample")       val sample:     Long,
    @Column("service")      val service:    String,
    @Column("serial")       val serial:     String
) : Persistable<Work.Companion.WorkPK> {
    @Id @Transient lateinit var _id:        WorkPK
    companion object{
        data class WorkPK(
            val sheet:      UUID,
            val work:       Int,
            val sample:     Long,
            val service:    String
        )
    }

    override fun getId(): WorkPK {
        return WorkPK(sheet, worklist, sample, service)
    }

    override fun isNew(): Boolean {
        return false;
    }
}